std/internal/cstring.d

```diff
@@ -91,41 +101,86 @@ if(isSomeChar!To && isSomeChar!From)
         @disable this(this);
         alias ptr this;
 
-        @property inout(To)* buffPtr() inout @safe pure
-        { return _ptr == useStack ? _buff.ptr : _ptr; }
+        @property inout(To)* buffPtr() inout pure
+        {
+            return _ptr == useStack ? _buff.ptr : _ptr;
+        }
 
-        @property const(To)* ptr() const @safe pure
-        { return buffPtr; }
+        @property const(To)* ptr() const pure
+        {
+            return buffPtr;
+        }
 
         ~this()
-        { if(_ptr != useStack) rawFree(_ptr); }
+        {
+            if (_ptr != useStack)
+            {
+                import core.stdc.stdlib : free;
+                free(_ptr);
+            }
+        }
 
     private:
         To* _ptr;
-        To[256] _buff;
-    }
-
-    // TODO: Don't stack allocate uninitialized array to
-    // not confuse unprecise GC.
+        version (unittest)
+        {
+            enum buffLength = 16 / To.sizeof;   // smaller size to trigger reallocations
+        }
+        else
+        {
+            enum buffLength = 256 / To.sizeof;   // production size
+        }
 
-    Res res = void;
-    if(!str.ptr)
-    {
-        res._ptr = null;
-        return res;
+        To[256 / To.sizeof] _buff;  // the 'small string optimization'
     }
 
+    Res res = void;     // expensive to fill _buff[]
+
     // Note: res._ptr can't point to res._buff as structs are movable.
 
-    bool overflow = false;
-    const totalCount = addu(maxLength!To(str), 1, overflow);
-    if(overflow)
-        onOutOfMemoryError();
-    const needAllocate = totalCount > res._buff.length;
-    To[] arr = copyEncoded(str, needAllocate ?
-        allocate!To(totalCount)[0 .. $ - 1] : res._buff[0 .. totalCount - 1]);
-    *(arr.ptr + arr.length) = '\0';
-    res._ptr = needAllocate ? arr.ptr : useStack;
+    import std.utf : byUTF;
+
+    To* p      = res._buff.ptr;
+    size_t len = res.buffLength;
+    size_t i;
+
+    static if (isSomeString!From)
+        auto r = cast(const(CF)[])str;
```

Comment:
Ah, right. inout has many issues with generic code, we need to work on a solution for that.